### PR TITLE
User auth tests

### PR DESCRIPTION
**Pull request by carystokes**

### CODE REVIEW

**Comment by carystokes** [comment]
Adds user functionality including registration and log-in.

File: spec/factories/views.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :view do
+    
+  end
+end
```
Comment (jennceng) [inline spec/factories/views.rb]: empty factory?

File: db/migrate/20161003194407_devise_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
+      # t.string   :unlock_token # Only if unlock strategy is :email or :both
+      # t.datetime :locked_at
+
+
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/migrate/20161003194407_devise_create_users.rb:35]: This was an automatic devise file and within a migration.

File: db/migrate/20161003194407_devise_create_users.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      t.inet     :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/migrate/20161003194407_devise_create_users.rb:31]: Also in a migration

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+RSpec.describe User, type: :model do
+  it { should have_valid(:first_name).when('John', 'Sally') }
+  it { should_not have_valid(:first_name).when(nil, '') }
+
+  it { should have_valid(:last_name).when('Smith', 'Swanson') }
+  it { should_not have_valid(:last_name).when(nil, '') }
+
+  it { should have_valid(:email).when('user@example.com', 'another@gmail.com') }
+  it { should_not have_valid(:email).when(nil, '', 'user', 'users@com', 'usersba.com') }
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/models/user_spec.rb:11]: Fixed

File: db/migrate/20161003194407_devise_create_users.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
+class DeviseCreateUsers < ActiveRecord::Migration
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/migrate/20161003194407_devise_create_users.rb:1]: Migration

File: db/migrate/20161003194407_devise_create_users.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  def change
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/migrate/20161003194407_devise_create_users.rb:2]: Migration

File: db/migrate/20161003194407_devise_create_users.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :first_name, null: false
+      t.string :last_name, null: false
+      t.string :email,              null: false, default: ""
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/migrate/20161003194407_devise_create_users.rb:7]: Migration

File: db/migrate/20161003194407_devise_create_users.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      t.string :encrypted_password, null: false, default: ""
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/migrate/20161003194407_devise_create_users.rb:8]: Migration

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/models/user_spec.rb:1]: Added '# frozen_string_literal: true'

File: spec/factories/users.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+
+
+FactoryGirl.define do
+
+
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/factories/users.rb]: Fixed

File: spec/factories/users.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+FactoryGirl.define do
+
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/factories/users.rb]: Fixed

File: spec/features/user_registration_spec.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    scenario 'user already logged in' do
+      visit new_user_registration_path
+      fill_in 'First name', with: 'Jon'
+      fill_in 'Last name', with: 'Smith'
+      fill_in 'Email', with: 'user@example.com'
+      fill_in 'Password', with: 'password'
+      fill_in 'Password confirmation', with: 'password'
+      click_button 'Sign Up'
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/features/user_registration_spec.rb:30]: Not found

File: spec/features/user_registration_spec.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  context 'as a user' do
+    scenario 'user provides valid user information for registration' do
+      visit new_user_registration_path
+      fill_in 'First name', with: 'Jon'
+      fill_in 'Last name', with: 'Smith'
+      fill_in 'Email', with: 'user@example.com'
+      fill_in 'Password', with: 'password'
+      fill_in 'Password confirmation', with: 'password'
+      click_button 'Sign Up'
+      expect(page).to have_content("Welcome! You have signed up successfully.")
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/features/user_registration_spec.rb:14]: Fixed

File: spec/features/user_registration_spec.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
+require 'rails_helper'
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/features/user_registration_spec.rb:1]: Fixed

File: spec/features/user_log_in_spec.rb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+      click_button 'Sign Up'
+      click_link 'Log Off'
+      expect(page).to have_content("Signed out successfully.")
+
+      visit new_user_session_path
+      fill_in 'Email', with: 'user@example.com'
+      fill_in 'Password', with: 'password'
+
+      click_button 'Log in'
+      expect(page).to have_content("Signed in successfully.")
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply spec/features/user_log_in_spec.rb:14]: Fixed

File: db/schema.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    t.datetime "last_sign_in_at"
+    t.inet     "current_sign_in_ip"
+    t.inet     "last_sign_in_ip"
+    t.datetime "created_at",                          null: false
+    t.datetime "updated_at",                          null: false
+  end
+
+  add_index "users", ["email"], name: "index_users_on_email", unique: true, using: :btree
+  add_index "users", ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true, using: :btree
+
```
[the remark this answers is not in the document]
Comment (larisabainton) [reply db/schema.rb:38]: Can't change schema

**Pull request merged**